### ISSUE TITLE
Segfault in R package with repeated model fits

### ISSUE THREAD

**Issue by rgranvil** [Microsoft/LightGBM#1468]
In the R package I get seemingly random segfaults when doing repeated fits of lightgbm using different feature subsets with num_threads > 1.  I've been running into this issue sporadically for several months now but it's been hard to isolate the issue since it seems to occur randomly.

## Environment info
Operating System: Red Hat Enterprise Linux
CPU: Intel(R) Xeon(R) CPU           X5677  @ 3.47GHz
C++/Python/R version: R version 3.4.2

## Error Message:
 *** caught segfault ***
address 0xa1c, cause 'memory not mapped'

Traceback:
 1: lgb.call("LGBM_DatasetCreateFromCSC_R", ret = handle, private$raw_data@p,     private$raw_data@i, private$raw_data@x, length(private$raw_data@p),     length(private$raw_data@x), nrow(private$raw_data), params_str,     ref_handle)
 2: data$construct()
 3: lgb.train(data = dtrain, valids = list(trn = dtrain), nrounds = 1,     params = params)
An irrecoverable exception occurred. R is aborting now ...

## Reproducible examples
This seems to happen somewhat randomly when doing repeated fits of lightgbm.  This is most reproducible script I've been able to make.  It seems to fail on a random iteration each time.

```r
library(lightgbm)
library(Matrix)
data(agaricus.train, package = "lightgbm")
for(i in 1:100000) {
  cat("Running step: ", i)
  
  # Randomly sample features to include
  all_features <- ncol(agaricus.train$data)
  selected_features <- sample((5:all_features), 1)
  features <- sample(1:all_features, selected_features)
  
  dtrain <- lgb.Dataset(agaricus.train$data[, features], label = agaricus.train$label)
  params <- list(objective = "regression", metric = "l2", num_threads = 4)
  lgb <- lgb.train(
    data = dtrain
    ,valids = list(trn = dtrain)
    ,nrounds = 1
    ,params = params
  )
}
```

**Comment by guolinke** [comment]
It seems a R issue:
http://r.789695.n4.nabble.com/What-is-the-meaning-of-quot-segfault-memory-not-mapped-quot-td870869.html 
https://stackoverflow.com/questions/17933445/caught-segfault-memory-not-mapped-error

@Laurae2 any ideas ?

**Comment by rgranvil** [comment]
also linking a similar xgboost issue from a few years ago if it helps:
https://github.com/dmlc/xgboost/issues/717

**Comment by Laurae2** [comment]
@rgranvil @guolinke Could not reproduce the results unfortunately even with 1 million.

Operating System: Windows Server 2016, Clear Linux latest version
CPU: i7-7700HQ, Dual Xeon Platinum 8180
C++/Python/R version: R version 3.5, LightGBM master

Note that this error might appear depending on the IDE you are running code. I know that R, Rgui, RStudio Desktop, and RStudio Server have different behaviors (I'm using RStudio Desktop, but RStudio Server would be likely severely prone to this issue).

The error seems that training occurs before the data is set in memory. Could you try the script below?

My script below goes to 1 million models, forces sleeping before running LightGBM, and doesn't show the training text:

```r
library(lightgbm)
library(Matrix)
data(agaricus.train, package = "lightgbm")

for (i in 1:1000000) {
  cat("\rRunning step: ", i)
  
  # Randomly sample features to include
  all_features <- ncol(agaricus.train$data)
  selected_features <- sample((5:all_features), 1)
  features <- sample(1:all_features, selected_features)
  
  dtrain <- lgb.Dataset(agaricus.train$data[, features], label = agaricus.train$label)
  Sys.sleep(0) # This is roughly 250ms for 1 million iterations (= 250 nanoseconds)
  
  params <- list(objective = "regression", metric = "l2", num_threads = 4)
  lgb <- lgb.train(
    data = dtrain,
    valids = list(trn = dtrain),
    nrounds = 1,
    params = params,
    verbose = -1
  )

}
```

**Comment by rgranvil** [comment]
@Laurae2 @guolinke Thanks for testing.  Adding Sys.sleep(0) did not fix the issue, but I tried updating to master (I had been using v2.1.1) and I've not come across the issue yet.  I went back and confirmed I could reproduce the error with a fresh install of 2.1.1, so it looks like maybe something was fixed between 2.1.1 and master?  Sorry I should have tried this earlier, but for now the issue appears to be resolved.
